### PR TITLE
Improving cache lookup to reduce recomputing / searches

### PR DESCRIPTION
**Pull request by mjmbischoff**
Adressing:
`// intentionally non-locking for simplicity...it's OK if we re-put the same key/value in the cache during a race condition.`

- Avoiding race condition while keeping cache lookup fast / non-blocking using a combination of `computeIfAbsent(..)` and `CompletableFuture`
- Invalidation of the cache is made idempotent by using the original value to avoid invalidating again but is called for each in flight request
- Cache entries are slightly bigger due to bookkeeping but seems like a good trade-off

### CODE REVIEW

**Comment by martijnvg** [comment]
Hey @mjmbischoff, thanks for putting this PR up. However I wonder whether the current approach in the PR can still lead to blocking threads? If I understand the `Cache#computeIfAbsent(...)` method correctly then in the case when there is no entry for key and it needs to be loaded, the first thread attempts to load it. But other threads during this process will wait until that thread has completed loading/computing a value (see line 435 in `Cache.java` and else statement at line 431).

**Comment by mjmbischoff** [comment]
Yes the first thread will do work. Technically all threads 'block', because as you know any call blocks, just for a very short time. The thread 'winning' out will have a little bit more work to do, although the others are blocked / parked at that point in time. This is because the `computeIfAbsent(..)` guarantees that one wins / only one value is calculated under contention. But all calls should return an Value.

The computation however is only the matter of creating an `CompletableFuture` Object and then dispatching an async call which should be short (in the sense that ppl will call it non blocking) The computation finishes and the `CompletableFuture` is returned on which we register a listener. The listener is called by the thread setting the value on the `CompletableFuture` so that will be the thread on which the async call was dispatched. Which is long running / blocking anyway because a `EnrichCoordinatorProxyAction` is executed there.

TL:DR the thread on the threadpoll of the originClient is the one that does all the work the rest is callbacks and cheap calls. 

Feel free to make the EnrichCoordinatorProxyAction slow to assert this. Now that I think of it I guess I could write a test for this.

**Comment by mjmbischoff** [comment]
Ah also forgot to mention that this approach also allows to cache failures, which it now immediately invalidates but perhaps it makes sense to observe some time out to save resources.

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java [hunk 1]
```diff
@@ -79,6 +86,23 @@ private String getEnrichIndexKey(SearchRequest searchRequest) {
         return ia.getIndices().get(0).getIndex().getName();
     }
 
+    CompletableFuture<SearchResponse> computeIfAbsent(
```
Comment (martijnvg) [inline x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java:89]: What about adding this method instead of these two methods:

```
void computeIfAbsent(
        SearchRequest searchRequest,
        BiConsumer<SearchRequest, ActionListener<SearchResponse>> executeSearch,
        BiConsumer<SearchResponse, Exception> handler
    ) {
        String enrichIndex = getEnrichIndexKey(searchRequest);
        CacheKey cacheKey = new CacheKey(enrichIndex, searchRequest);

        try {
            CompletableFuture<SearchResponse> cacheEntry = cache.computeIfAbsent(cacheKey, request -> {
                CompletableFuture<SearchResponse> completableFuture = new CompletableFuture<>();
                ActionListener<SearchResponse> listener = ActionListener.wrap(
                    completableFuture::complete,
                    completableFuture::completeExceptionally
                );
                executeSearch.accept(searchRequest, listener);
                return completableFuture;
            });
            cacheEntry.whenComplete((response, throwable) -> {
                Exception exception = null;
                if (throwable != null) {
                    cache.invalidate(cacheKey, cacheEntry);
                    if (throwable instanceof Exception) {
                        exception = (Exception) throwable;
                    } else {
                        exception = new Exception("Unexpected error", throwable);
                    }
                }
                handler.accept(response, exception);
            });
        } catch (ExecutionException e) {
            handler.accept(null, e);
        }
    }
```

and then `EnrichProcessorFactory#createSearchRunner()` can look like this:

```
Client originClient = new OriginSettingClient(client, ENRICH_ORIGIN);
        return (req, handler) -> {
            enrichCache.computeIfAbsent(
                req,
                (searchRequest, listener) -> originClient.execute(EnrichCoordinatorProxyAction.INSTANCE, searchRequest, listener),
                handler
            );
        };
```

This way the logic of dealing with `Futures` is kept inside the cache and then there is also no need to expose the `invalidate(...)` method.

Comment (martijnvg) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java:89]: (small nit: maybe move this method up to the other methods with package protected visibility)

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java [hunk 2]
```diff
@@ -37,21 +40,25 @@
...
-    SearchResponse get(SearchRequest searchRequest) {
+    CompletableFuture<SearchResponse> get(SearchRequest searchRequest) {
         String enrichIndex = getEnrichIndexKey(searchRequest);
         CacheKey cacheKey = new CacheKey(enrichIndex, searchRequest);
 
         return cache.get(cacheKey);
     }
 
     void put(SearchRequest searchRequest, SearchResponse searchResponse) {
+        put(searchRequest, CompletableFuture.completedFuture(searchResponse));
```
Comment (martijnvg) [inline x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java:58]: These put methods can be removed and the `EnrichCacheTests` test needs to be converted to use `computeIfAbsent(...)` method.

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichProcessorFactory.java [hunk 3]
```diff
@@ -129,15 +131,30 @@ public void accept(ClusterState state) {
...
+                        ActionListener.wrap(completableFuture::complete, completableFuture::completeExceptionally)
+                    );
+                    return completableFuture;
+                });
+                cacheEntry.whenComplete((response, throwable) -> {
+                    Exception exception = null;
+                    if (throwable != null) {
+                        enrichCache.invalidate(req, cacheEntry);
+                        if (throwable instanceof Exception) {
+                            exception = (Exception) throwable;
```
Comment (martijnvg) [inline x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichProcessorFactory.java:149]: In case throwable is an instance of `Error` then we should throw it (`throw (Error) throwable;`).
So that the `ElasticsearchUncaughtExceptionHandler` can catch it.

**Review by martijnvg — COMMENTED** [review]
Thanks for explaining @mjmbischoff. I left a few comments.

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java (hunk 1 above)
Comment (mjmbischoff) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java:89]: Pulling the logic up makes sense, will give it a go.

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichProcessorFactory.java (hunk 3 above)
Comment (mjmbischoff) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichProcessorFactory.java:149]: Ah yes, a little bit of laziness on my side wasn't sure what was up the call chain and what the implications would be.

Comment (mjmbischoff) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichProcessorFactory.java:149]: Hmm this might violate the call back contract - if we throw the error we don't call the callback which might mean we'll hold onto / leak resources. I would make it the callers and the search logic responsible for handling this?

I could of course call the callback and then throw but then we might end up throwing the error twice, if I wrap it and otherwise we lose context. Both don't sound too appealing.

Comment (martijnvg) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichProcessorFactory.java:149]: `ElasticsearchUncaughtExceptionHandler` will halt the node if a `Error` is encountered. So invoked a callback isn't important, since the jvm will stop. So I think in this case breaking contract is okay.

Wrapping a jvm error into another exception hides potentially bad errors and leave the jvm in a bad state, which is why the in the ES code base `Throwable` is never caught (just `Exception`), so that jvm errors end up in  `ElasticsearchUncaughtExceptionHandler `.

Comment (mjmbischoff) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichProcessorFactory.java]: Ah excellent. Yeah all bets are off on an error so killing the node is the best we can do. I'll update the code accordingly.

Comment (mjmbischoff) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichProcessorFactory.java]: Hmm A bit of an edge case, but upon looking at `ElasticsearchUncaughtExceptionHandler` it seems we still deem some throwables as non-fatal meaning we keep the instance around. Opted to report a shallow exception to the callback and then throw the throwable for `ElasticsearchUncaughtExceptionHandler` to handle. If it exits then the callback doesn't matter if the process limps on we try not to make it worse.

**Comment by mjmbischoff** [comment]
Think we're getting close. Added comments to the code to clarify (or to make my miss-assumptions explicit. ;-))

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichProcessorFactory.java (hunk 3 above)
Comment (mjmbischoff) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichProcessorFactory.java]: Addressed in https://github.com/mjmbischoff/elasticsearch/pull/77259/commits/5b9d9bc73cb116248560a036b0fd6559fef85d96

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java [hunk 4]
```diff
@@ -73,12 +78,60 @@ void setMetadata(Metadata metadata) {
...
+            cacheEntry.whenComplete((response, throwable) -> {
+                if (throwable != null) {
+                    scheduleClearFailure(cacheKey, cacheEntry);
+                    if (throwable instanceof Exception) {
+                        callBack.accept(response, (Exception) throwable);
+                        return;
+                    } else {
+                        // Let ElasticsearchUncaughtExceptionHandler handle this, should it deem it non-fatal let's still trigger the
+                        // callback to ensure proper clean up.
+                        callBack.accept(response, new Exception("Throwable was raised '" + throwable.getMessage() + "'"));
```
Comment (martijnvg) [inline x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java:113]: In this case `throwable` is an error and `ElasticsearchUncaughtExceptionHandler` always halts the process if an uncaught exception is of type error. So I think this else clause should just throw the error.

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java [hunk 5]
```diff
@@ -73,12 +78,60 @@ void setMetadata(Metadata metadata) {
...
+    }
+
     private String getEnrichIndexKey(SearchRequest searchRequest) {
         String alias = searchRequest.indices()[0];
         IndexAbstraction ia = metadata.getIndicesLookup().get(alias);
         return ia.getIndices().get(0).getIndex().getName();
     }
 
+    private void scheduleClearFailure(CacheKey cacheKey, CompletableFuture<SearchResponse> cacheEntry) {
+        // For now, we schedule immediately. But perhaps we find a reason to cache failures as well.
```
Comment (martijnvg) [inline x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java:131]: If an error occurs searching an enrich index then I don't think this should be cached. The search requests are controlled by the processor / policy type and not by the user. So an exception should be seen as a bug.

File: x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichCacheTests.java [hunk 6]
```diff
@@ -89,13 +95,14 @@ public void testCaching() {
...
-        enrichCache.put(searchRequest3, searchResponse);
+        // warming the cache means there's a miss
+        warmCache(enrichCache, searchRequest1, searchResponse);
+        warmCache(enrichCache, searchRequest2, searchResponse);
+        warmCache(enrichCache, searchRequest3, searchResponse);
         var cacheStats = enrichCache.getStats("_id");
         assertThat(cacheStats.getCount(), equalTo(3L));
         assertThat(cacheStats.getHits(), equalTo(0L));
-        assertThat(cacheStats.getMisses(), equalTo(0L));
+        assertThat(cacheStats.getMisses(), equalTo(3L));
```
Comment (martijnvg) [inline x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichCacheTests.java:105]: why is the expected miscount increased here?

**Review by martijnvg — COMMENTED** [review]
Thanks @mjmbischoff for the update. I left a few comments.

**Comment by martijnvg** [comment]
@elasticmachine update branch

File: x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichCacheTests.java (hunk 6 above)
Comment (mjmbischoff) [reply x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichCacheTests.java:105]: Without introducing a method specifically for the tests, warming the cache requires a cache miss to trigger a resolve.

The class is also final so can't do the usual trickery to break encapsulation. A specific put or constructor where we can pass in a `Cache<CacheKey, CompletableFuture<SearchResponse>> cache` are alternatives - I don't know whats preferred?

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java (hunk 4 above)
Comment (mjmbischoff) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java]: Ok, assuming the non-Exception Throwable is always an Error. Addressed in https://github.com/elastic/elasticsearch/pull/77259/commits/c38f0970ecc6eeadca47f52162eb3aa6263c06c5

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java (hunk 5 above)
Comment (mjmbischoff) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java]: I've removed the premature indirection and inlined the invalidation. https://github.com/elastic/elasticsearch/pull/77259/commits/c38f0970ecc6eeadca47f52162eb3aa6263c06c5

File: x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichCacheTests.java (hunk 6 above)
Comment (martijnvg) [reply x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichCacheTests.java:105]: I see. Maybe we should make `EnrichCache` non final?

Comment (mjmbischoff) [reply x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichCacheTests.java]: Adressed in https://github.com/elastic/elasticsearch/pull/77259/commits/14ef0a2eb130895d04086722545ca5c7ea1cc715

**Review by martijnvg — APPROVED** [review]
LGTM 👍

**Comment by mjmbischoff** [comment]
@martijnvg As always many thanks for reviewing! Looking at https://github.com/elastic/elasticsearch/pull/76800 I think it's targeted for 7.16 but I'm unsure if that train has already left the station, can I leave the backporting, if any, to you?

**Comment by martijnvg** [comment]
@mjmbischoff I will backport this to 7.16, there is still some time before feature freeze.
